src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala

```diff
@@ -1189,13 +1189,13 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
 
   private def unify(tp1: List[Type], tp2: List[Type], env: TypeEnv, strict: Boolean): TypeEnv = {
     if (tp1.isEmpty || tp2.isEmpty) env
-    else (tp1 zip tp2).foldLeft(env) { (env, args) =>
-      if (!strict) unify(args._1, args._2, env, strict)
+    else foldLeft2(tp1, tp2)(env) { (env, arg1, arg2) =>
+      if (!strict) unify(arg1, arg2, env, strict)
       else {
-        val nenv = unify(args._1, args._2, emptyEnv, strict)
+        val nenv = unify(arg1, arg2, emptyEnv, strict)
         if (env.keySet.intersect(nenv.keySet).isEmpty) env ++ nenv
```

Comment:
`env.keySet.forall(!nenv.contains)`? I know it's off topic, but...

I haven't actually measured if that's noticeably better, mind; I have no idea how big `env`s tend to be.

If you want to try other things for that, you could also consider unconditionally making the set union, and then checking if its size is the sum of its constituents'. (I forget how expensive `size` is.)

As usual, if `TypeEnv`s tend to be small in normal code, I wouldn't worry about it.

Reply:
Given that `nenv` is usually the smallest of the two, perhaps iterating over its keys (without allocating any set) could at least save more allocations.

Reply:
In any case, since I am not too familiar yet with all the implementations of the Scala collections, I am not confident enough to apply your suggestion.